parlai/agents/vsepp_caption/modules.py

```diff
@@ -0,0 +1,215 @@
+        self.margin = margin
+        self.sim = cosine_sim
+        self.max_violation = max_violation
+
+    def forward(self, im, caps, offset=0):
+        # Compute the similarity of each image/caption pair
+        scores = self.sim(im, caps)
+        diagonal = scores.diag().view(im.shape[0], 1)
+        d1 = diagonal.expand(scores.size())
+        d2 = diagonal.t().expand(scores.size())
+
+        # Caption retrieval score
+        cost_cap = (self.margin + scores - d1).clamp(min=0)
+        # image retrieval score
+        cost_im = (self.margin + scores - d2).clamp(min=0)
+
+        mask = torch.eye(im.shape[0]) > 0.5
+        if self.use_cuda:
+            mask = mask.cuda()
+        cost_cap = cost_cap.masked_fill(mask, 0)
+        cost_im = cost_im.masked_fill(mask, 0)
+
+        # Compute the metrics (ranks, top1)
+        if self.use_cuda:
+            sorted_ranks = np.flip(np.argsort(scores.detach().cpu().numpy()), 1)
+        else:
+            sorted_ranks = np.flip(np.argsort(scores.detach().numpy()), 1)
+        top1 = sorted_ranks[:, 0]
+        ranks = []
+        for idx in range(im.shape[0]):
+            ranks.append(np.where(sorted_ranks[idx, :]==(idx + offset))[0][0])
+
+        # keep the maximum violating negative for each query
+        if self.max_violation:
+            cost_cap = cost_cap.max(1)[0]
+            cost_im = cost_im.max(0)[0]
+
+        return cost_cap.sum() + cost_im.sum(), ranks, top1
+
+
+class EncoderImage(nn.Module):
+    def __init__(self, embed_size, finetune=False, cnn_type='resnet152',
+                 no_imgnorm=False):
+        """Load pretrained CNN and replace top fc layer."""
+        super().__init__()
+        self.embed_size = embed_size
+        self.no_imgnorm = no_imgnorm
+
+        # Load a pre-trained model
+        self.cnn = self.get_cnn(cnn_type)
+
+        # For efficient memory usage.
+        for param in self.cnn.parameters():
+            param.requires_grad = finetune
+
+        # Replace the last fully connected layer of CNN with a new one
+        if cnn_type.startswith('vgg'):
+            self.fc = nn.Linear(self.cnn.classifier._modules['6'].in_features,
+                                embed_size)
+            self.cnn.classifier = nn.Sequential(
+                *list(self.cnn.classifier.children())[:-1])
+        elif cnn_type.startswith('resnet'):
+            self.fc = nn.Linear(self.cnn.module.fc.in_features, embed_size)
+            self.cnn.module.fc = nn.Sequential()
+
+        self.init_weights()
+
+    def get_cnn(self, arch):
+        """Load a pretrained CNN and parallelize over GPUs
+        """
+        print("=> using pre-trained model '{}'".format(arch))
+        model = models.__dict__[arch](pretrained=True)
+
+        if arch.startswith('alexnet') or arch.startswith('vgg'):
+            model.features = nn.DataParallel(model.features)
+            model.cuda()
+        else:
+            model = nn.DataParallel(model).cuda()
```

Comment:
shouldn't need to call cuda here--it should do it automatically if you call cuda on the encoder (and this forces cuda on)